docs/proposals/cluster-groups.md

```diff
@@ -0,0 +1,25 @@
+# Cluster Groups
+
+## Problem
+To allow cluster-scoped group definitions to be shared across multiple related projects, we'll introduce the concept of Groups (cluster-scoped sets of users) that are managed by cluster-admins.
+
+
+## Use Cases
+ 1.  I want to be able to define some Groups that will affect policy evaluation across all projects.
+ 1.  I want to be able to efficiently manage group membership.
+
+
+## Groups
+Groups are cluster-scoped sets of users that are managed by cluster-admins and may contain any user.  Groups can be bound to roles in all projects.  That means that if UserA is added to GroupB and ProjectC has a rolebinding that says "GroupB is an editor", UserA now has editor permissions in ProjectC.  Groups will be described using an API object that looks like this:
+```
+// Group contains the set of users that belong to a given Group.  The name may not contain ":"
+type Group struct {
+	kapi.TypeMeta
+	kapi.ObjectMeta
+
+	// Users is a list of Users that are members of the Group
+	Users util.StringSet
+}
+```
+
+A reverse lookup index will be built in memory (similar to the project cache) to be allow the authenticator to make fast lookups for Group membership determination.
```

Comment:
Group names and usernames in ldap can get pretty long.  1000 users is low
end, but high end is going to be something like 20k

On Mon, Jul 27, 2015 at 2:46 PM, David Eads <notifications@github.com>
wrote:

> In docs/proposals/cluster-groups.md
> <https://github.com/openshift/origin/pull/3753#discussion_r35571403>:
>
> > +
> > +
> > +## Groups
> > +Groups are cluster-scoped sets of users that are managed by cluster-admins and may contain any user.  Groups can be bound to roles in all projects.  That means that if UserA is added to GroupB and ProjectC has a rolebinding that says "GroupB is an editor", UserA now has editor permissions in ProjectC.  Groups will be described using an API object that looks like this:
> > +```
> > +// Group contains the set of users that belong to a given Group.  The name may not contain ":"
> > +type Group struct {
> > +	kapi.TypeMeta
> > +	kapi.ObjectMeta
> > +
> > +	// Users is a list of Users that are members of the Group
> > +	Users util.StringSet
> > +}
> > +```
> > +
> > +A reverse lookup index will be built in memory (similar to the project cache) to be allow the authenticator to make fast lookups for Group membership determination.
>
>  What's the numbers on the size of the reverse index here? If I have 1000
> users who are each members of 50 groups in ldap, is the fan out manageable?
> What about larger scales?
>
> For a while.
>
> count(users) * len(username) * count(avg group memberships) * len(groups)
>
> If we take your example: say 1024 * 32 * 64 * 32, we come out with 64MB.
> It's significant, but probably acceptable.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3753/files#r35571403>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift